rpcs/src/client/mock_rpc_client.rs

```diff
@@ -0,0 +1,643 @@
+// Copyright 2019-2025 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+//! This module exposes a [`MockRpcClient`], which is useful for testing.
+//! 
+//! # Example
+//! 
+//! ```rust
+//! use subxt_rpcs::client::{ RpcClient, MockRpcClient };
+//! use subxt_rpcs::client::mock_rpc_client::Json;
+//! 
+//! let mut state = vec![
+//!     Json(1u8),
+//!     Json(2u8),
+//!     Json(3u8),
+//! ];
+//! 
+//! // Define a mock client by providing some functions which intercept 
+//! // method and subscription calls and return some response.
+//! let mock_client = MockRpcClient::builder()
+//!     .method_handler_once("foo", move |params| {
+//!         // Return each item from our state, and then null afterwards.
+//!         let val = state.pop();
+//!         async move { val }
+//!     })
+//!     .subscription_handler("bar", |params, unsub| async move {
+//!         // Arrays, vecs or an RpcSubscription can be returned here to
+//!         // signal the set of values to be handed back on a subscription.
+//!         vec![Json(1), Json(2), Json(3)]
+//!     })
+//!     .build();
+//! 
+//! // Build an RPC Client that can be used in Subxt or in conjunction with
+//! // the RPC methods provided in this crate. 
+//! let rpc_client = RpcClient::new(mock_client);
+//! ```
+
+use super::{RpcClientT, RawRpcFuture, RawRpcSubscription};
+use crate::{Error, UserError};
+use core::future::Future;
+use futures::StreamExt;
+use serde_json::value::RawValue;
+use std::sync::{Arc, Mutex};
+use std::collections::{HashMap, VecDeque};
+
+type MethodHandlerFnOnce = Box<dyn FnOnce(&str, Option<Box<serde_json::value::RawValue>>) -> RawRpcFuture<'static, Box<RawValue>> + Send + Sync + 'static>;
+type SubscriptionHandlerFnOnce = Box<dyn FnOnce(&str, Option<Box<serde_json::value::RawValue>>, &str) -> RawRpcFuture<'static, RawRpcSubscription> + Send + Sync + 'static>;
+
+type MethodHandlerFn = Box<dyn FnMut(&str, Option<Box<serde_json::value::RawValue>>) -> RawRpcFuture<'static, Box<RawValue>> + Send + Sync + 'static>;
+type SubscriptionHandlerFn = Box<dyn FnMut(&str, Option<Box<serde_json::value::RawValue>>, &str) -> RawRpcFuture<'static, RawRpcSubscription> + Send + Sync + 'static>;
+
+/// A builder to configure and build a new [`MockRpcClient`].
+pub struct MockRpcClientBuilder {
+    method_handlers_once: HashMap<String, VecDeque<MethodHandlerFnOnce>>,
+    method_handlers: HashMap<String, MethodHandlerFn>,
+    method_fallback: Option<MethodHandlerFn>,
+    subscription_handlers_once: HashMap<String, VecDeque<SubscriptionHandlerFnOnce>>,
+    subscription_handlers: HashMap<String, SubscriptionHandlerFn>,
+    subscription_fallback: Option<SubscriptionHandlerFn>
+}
+
+impl  MockRpcClientBuilder {
```

Comment:
```suggestion
impl MockRpcClientBuilder {
```

Reply:
Weird that fmt did this or didn't catch it!